pkg/cloudprovider/aws/amifamily/ubuntu.go

```diff
@@ -0,0 +1,100 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package amifamily
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/cloudprovider/aws/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/utils/functional"
+	"github.com/aws/karpenter/pkg/utils/injection"
+)
+
+func init() {
+	ubuntu := Ubuntu{}
+	amiFamilies[ubuntu.Name()] = ubuntu
+}
+
+type Ubuntu struct{}
+
+// Name is the name of the AMI Family
+func (u Ubuntu) Name() string {
+	return v1alpha1.AMIFamilyUbuntu
+}
+
+// SSMAlias returns the AMI Alias to query SSM
+func (u Ubuntu) SSMAlias(version string, instanceType cloudprovider.InstanceType) string {
+	return fmt.Sprintf("/aws/service/canonical/ubuntu/eks/20.04/%s/stable/current/%s/hvm/ebs-gp2/ami-id", version, instanceType.Architecture())
+}
+
+// UserData returns the default userdata script for the AMI Family
+func (u Ubuntu) UserData(ctx context.Context, constraints *v1alpha1.Constraints, instanceTypes []cloudprovider.InstanceType, additionalLabels map[string]string, caBundle *string) string {
+	var containerRuntimeArg string
+	if !needsDocker(instanceTypes) {
+		containerRuntimeArg = "--container-runtime containerd"
+	}
+
+	var userData bytes.Buffer
+	userData.WriteString(fmt.Sprintf(`#!/bin/bash -xe
+exec > >(tee /var/log/user-data.log|logger -t user-data -s 2>/dev/console) 2>&1
+/etc/eks/bootstrap.sh '%s' %s \
+    --apiserver-endpoint '%s'`,
+		injection.GetOptions(ctx).ClusterName,
+		containerRuntimeArg,
+		injection.GetOptions(ctx).ClusterEndpoint))
+	if caBundle != nil {
+		userData.WriteString(fmt.Sprintf(` \
+    --b64-cluster-ca '%s'`,
+			*caBundle))
+	}
+
+	nodeLabelArgs := getNodeLabelArgs(functional.UnionStringMaps(additionalLabels, constraints.Labels))
+	nodeTaintsArgs := getNodeTaintArgs(constraints)
+	kubeletExtraArgs := strings.Trim(strings.Join([]string{nodeLabelArgs, nodeTaintsArgs.String()}, " "), " ")
+
+	if !injection.GetOptions(ctx).AWSENILimitedPodDensity {
+		userData.WriteString(` \
+    --use-max-pods=false`)
+		kubeletExtraArgs += " --max-pods=110"
+	}
+
+	if len(kubeletExtraArgs) > 0 {
+		userData.WriteString(fmt.Sprintf(` \
+    --kubelet-extra-args '%s'`, kubeletExtraArgs))
+	}
+	if len(constraints.KubeletConfiguration.ClusterDNS) > 0 {
+		userData.WriteString(fmt.Sprintf(` \
+    --dns-cluster-ip '%s'`, constraints.KubeletConfiguration.ClusterDNS[0]))
+	}
+	return base64.StdEncoding.EncodeToString(userData.Bytes())
+}
+
+// BlockDeviceMappings returns the default block device mappings for the AMI Family
+func (u Ubuntu) BlockDeviceMappings(constraints *v1alpha1.Constraints) []*v1alpha1.BlockDeviceMapping {
+	if len(constraints.BlockDeviceMappings) != 0 {
```

Comment:
Worth doing this check outside and renaming this function DefaultBlockDeviceMappings?